python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -275,6 +275,38 @@ def __iter__(self):
         for i, name in enumerate(group_names):
             yield name, grouped_values[offsets[i] : offsets[i + 1]]
 
+    @property
+    def dtypes(self):
+        """
+        Return the dtypes in this group.
+
+        Returns
+        -------
+        pandas.DataFrame
+            The data type of each column of the group.
+
+        Examples
+        --------
+        >>> import cudf
+        >>> df = cudf.DataFrame({'a': [1, 2, 3, 3], 'b': ['x', 'y', 'z', 'a'],
+        ...                      'c':[10, 11, 12, 12]})
+        >>> df.groupby("a").dtypes
+                b      c
+        a
+        1  object  int64
+        2  object  int64
+        3  object  int64
+        """
+        index = self.grouping.keys.unique().to_pandas()
+        result = pd.DataFrame(
+            {
+                name: pd.Series([self.obj._dtypes[name]]).repeat(len(index))
+                for name in self.grouping.values._column_names
+            },
+        )
+        result.index = index
+        return result
```

Comment:
Construct with the desired index, and directly return.
```suggestion
        return pd.DataFrame(
            {
                name: pd.Series([self.obj._dtypes[name]]).repeat(len(index))
                for name in self.grouping.values._column_names
            },
            index=index,
        )
```

Reply:
Like I discribed in:

We can't actually do that via constructor, via constructor the Series objects are always reindexed:

```ipython
In [5]: s = pd.Series([1]).repeat(3)

In [6]: s
Out[6]: 
0    1
0    1
0    1
dtype: int64

In [10]: pd.DataFrame({"a":s}, index=pd.Index(['a', 'b', 'c']))
<ipython-input-10-c45a4113f45b>:1: FutureWarning: reindexing with a non-unique Index is deprecated and will raise in a future version.
  pd.DataFrame({"a":s}, index=pd.Index(['a', 'b', 'c']))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[10], line 1
----> 1 pd.DataFrame({"a":s}, index=pd.Index(['a', 'b', 'c']))

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/frame.py:664, in DataFrame.__init__(self, data, index, columns, dtype, copy)
    658     mgr = self._init_mgr(
    659         data, axes={"index": index, "columns": columns}, dtype=dtype, copy=copy
    660     )
    662 elif isinstance(data, dict):
    663     # GH#38939 de facto copy defaults to False only in non-dict cases
--> 664     mgr = dict_to_mgr(data, index, columns, dtype=dtype, copy=copy, typ=manager)
    665 elif isinstance(data, ma.MaskedArray):
    666     import numpy.ma.mrecords as mrecords

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/internals/construction.py:493, in dict_to_mgr(data, index, columns, dtype, typ, copy)
    489     else:
    490         # dtype check to exclude e.g. range objects, scalars
    491         arrays = [x.copy() if hasattr(x, "dtype") else x for x in arrays]
--> 493 return arrays_to_mgr(arrays, columns, index, dtype=dtype, typ=typ, consolidate=copy)

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/internals/construction.py:123, in arrays_to_mgr(arrays, columns, index, dtype, verify_integrity, typ, consolidate)
    120         index = ensure_index(index)
    122     # don't force copy because getting jammed in an ndarray anyway
--> 123     arrays = _homogenize(arrays, index, dtype)
    124     # _homogenize ensures
    125     #  - all(len(x) == len(index) for x in arrays)
    126     #  - all(x.ndim == 1 for x in arrays)
   (...)
    129 
    130 else:
    131     index = ensure_index(index)

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/internals/construction.py:599, in _homogenize(data, index, dtype)
    595         val = val.astype(dtype, copy=False)
    596     if val.index is not index:
    597         # Forces alignment. No need to copy data since we
    598         # are putting it into an ndarray later
--> 599         val = val.reindex(index, copy=False)
    601     val = val._values
    602 else:

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/series.py:5094, in Series.reindex(self, *args, **kwargs)
   5090         raise TypeError(
   5091             "'index' passed as both positional and keyword argument"
   5092         )
   5093     kwargs.update({"index": index})
-> 5094 return super().reindex(**kwargs)

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/generic.py:5289, in NDFrame.reindex(self, *args, **kwargs)
   5286     return self._reindex_multi(axes, copy, fill_value)
   5288 # perform the reindex on the axes
-> 5289 return self._reindex_axes(
   5290     axes, level, limit, tolerance, method, fill_value, copy
   5291 ).__finalize__(self, method="reindex")

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/generic.py:5309, in NDFrame._reindex_axes(self, axes, level, limit, tolerance, method, fill_value, copy)
   5304 new_index, indexer = ax.reindex(
   5305     labels, level=level, limit=limit, tolerance=tolerance, method=method
   5306 )
   5308 axis = self._get_axis_number(a)
-> 5309 obj = obj._reindex_with_indexers(
   5310     {axis: [new_index, indexer]},
   5311     fill_value=fill_value,
   5312     copy=copy,
   5313     allow_dups=False,
   5314 )
   5315 # If we've made a copy once, no need to make another one
   5316 copy = False

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/generic.py:5355, in NDFrame._reindex_with_indexers(self, reindexers, fill_value, copy, allow_dups)
   5352     indexer = ensure_platform_int(indexer)
   5354 # TODO: speed up on homogeneous DataFrame objects (see _reindex_multi)
-> 5355 new_data = new_data.reindex_indexer(
   5356     index,
   5357     indexer,
   5358     axis=baxis,
   5359     fill_value=fill_value,
   5360     allow_dups=allow_dups,
   5361     copy=copy,
   5362 )
   5363 # If we've made a copy once, no need to make another one
   5364 copy = False

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/internals/managers.py:737, in BaseBlockManager.reindex_indexer(self, new_axis, indexer, axis, fill_value, allow_dups, copy, only_slice, use_na_proxy)
    735 # some axes don't allow reindexing with dups
    736 if not allow_dups:
--> 737     self.axes[axis]._validate_can_reindex(indexer)
    739 if axis >= self.ndim:
    740     raise IndexError("Requested axis not found in manager")

File /nvme/0/pgali/envs/cudfdev/lib/python3.10/site-packages/pandas/core/indexes/base.py:4316, in Index._validate_can_reindex(self, indexer)
   4314 # trying to reindex on an axis with duplicates
   4315 if not self._index_as_unique and len(indexer):
-> 4316     raise ValueError("cannot reindex on an axis with duplicate labels")

ValueError: cannot reindex on an axis with duplicate labels

In [11]: s.reset_index(drop=True)
Out[11]: 
0    1
1    1
2    1
dtype: int64

In [12]: pd.DataFrame({"a":s.reset_index(drop=True)}, index=pd.Index(['a', 'b', 'c']))
Out[12]: 
    a
a NaN
b NaN
c NaN

In [13]: df = pd.DataFrame({"a":s})

In [14]: df.index = pd.Index(['a', 'b', 'c'])

In [15]: df
Out[15]: 
   a
a  1
b  1
c  1
```

Reply:
Do the inputs really need to be `Series` though? Would something like this not work?


```python
        return pd.DataFrame(
            {
                name: [self.obj._dtypes[name]]*len(index)
                for name in self.grouping.values._column_names
            },
            index=index,
        )

```

Reply:
@shwina You are right, we don't need a Series constructor, I initially used it for `repeat`, but turns out it is slower too:
```python
In [10]: timeit.timeit(lambda: pd.Series(["int64"]).repeat(100_00_000), number=100)
Out[10]: 13.081491015851498

In [12]: timeit.timeit(lambda: ["int64"]*100_00_000, number=100)
Out[12]: 5.543706430122256
```